File: source/daemon/crates/wardnet-common/src/network_zone.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// A permitted routing-target *kind* a zone allows. Coarse: any tunnel, or direct.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, utoipa::ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum AllowedTargetKind {
    Direct,
    Tunnel,
}

/// Cross-zone isolation rung of the guarantee ladder (epic #244). Only the rungs
/// we have issues for exist; VLAN is a documented non-goal (ADR), not a variant.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, utoipa::ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum ZoneStance {
    /// nftables egress + admin-UI gating only; peer isolation delegated to the AP. (CI-2 #736)
    SharedSubnet,
    /// Per-device /32 + proxy-ARP; requires wardnet-DHCP-mode. (CI-3 #737)
    IsolateMembers,
}

/// Whether a zone was seeded by the daemon or created by an admin.
/// Mirrors the existing DNS "Zone provenance" (system | manual).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, utoipa::ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum ZoneProvenance {
    System,
    Manual,
}

/// A per-zone subnet (Phase 2 / DHCP-mode). Recorded-only in #735; always None on seeds.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, utoipa::ToSchema)]
pub struct ZoneSubnet {
    /// CIDR, e.g. "10.44.0.0/24". Validated when set; unused by any Phase-1 code path.
    pub cidr: String,
}

/// A Network Zone: a named policy bucket a device belongs to (exactly one) that
/// gates the device's allowed routing targets, its reachability of the Pi's
/// admin surfaces, and (Phase 2+) its network isolation. See epic #244 and the
/// `adr-network-zone-isolation` ADR.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, utoipa::ToSchema)]
#[allow(clippy::struct_excessive_bools)]
pub struct NetworkZone {
    pub id: Uuid,
    pub name: String,
    pub provenance: ZoneProvenance,
    pub isolation_stance: ZoneStance,
    /// Routing-target kinds a device in this zone may pick. Empty is rejected on write.
    pub allowed_targets: Vec<AllowedTargetKind>,
    /// Within an isolate-members zone, also isolate same-zone peers. (CI-3 #737)
    pub member_isolation: bool,
    /// Per-zone subnet (CI-3 #737). None for all seeds.
    pub subnet: Option<ZoneSubnet>,
    /// May devices in this zone reach the Pi's admin surfaces? (intent-only in Phase 1)
    pub admin_ui_reachable: bool,
    /// The protected "home"/anchor zone. Exactly one true. Deletion-guarded.
    pub is_default: bool,
    /// Where freshly-discovered devices are assigned. Exactly one true.
    pub is_default_for_new: bool,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl NetworkZone {
    /// Does this zone permit the given resolved concrete target kind?
    #[must_use]
    pub fn permits_kind(&self, kind: AllowedTargetKind) -> bool {
        self.allowed_targets.contains(&kind)
    }
}

File: source/sdk/wardnet-js/src/types/network-zone.ts
/**
 * Network Zones (epic #244, issue #735).
 *
 * A Network Zone is a named policy bucket a device belongs to (exactly one). It
 * gates the device's allowed routing targets, its reachability of the Pi's admin
 * surfaces, and (Phase 2+) its network isolation. Distinct from the DNS
 * *authoritative local zone* (`types/dns-local`).
 */

/** A permitted routing-target kind. Coarse: any tunnel, or direct. */
export type AllowedTargetKind = "direct" | "tunnel";

/** Cross-zone isolation rung of the guarantee ladder. VLAN is a non-goal. */
export type ZoneStance = "shared_subnet" | "isolate_members";

/** Whether a zone was seeded by the daemon or created by an admin. */
export type ZoneProvenance = "system" | "manual";

/** A per-zone subnet (Phase 2 / DHCP-mode). Recorded-only in Phase 1. */
export interface ZoneSubnet {
  /** CIDR, e.g. "10.44.0.0/24". */
  cidr: string;
}

/** A Network Zone. */
export interface NetworkZone {
  id: string;
  name: string;
  provenance: ZoneProvenance;
  isolation_stance: ZoneStance;
  /** Routing-target kinds a device in this zone may pick. Never empty. */
  allowed_targets: AllowedTargetKind[];
  /** Within an isolate-members zone, also isolate same-zone peers. */
  member_isolation: boolean;
  /** Per-zone subnet. `null` for all seeds. */
  subnet: ZoneSubnet | null;
  /** May devices in this zone reach the Pi's admin surfaces? */
  admin_ui_reachable: boolean;
  /** The protected "home"/anchor zone. Exactly one true. Deletion-guarded. */
  is_default: boolean;
  /** Where freshly-discovered devices are assigned. Exactly one true. */
  is_default_for_new: boolean;
  created_at: string;
  updated_at: string;
}

/** A zone plus its current member count (list/detail view). */
export interface NetworkZoneView extends NetworkZone {
  member_count: number;
}

/**
 * Minimal zone info the daemon attaches to `GET /api/devices/me` for the
 * read-only zone display in the user PWA. The caller is device-keyed and can't
 * read the admin-only zone catalog, so this is the only zone data it sees.
 */
export interface ZoneSummary {
  /** Human-readable zone name (e.g. "Guest"). */
  name: string;
  /** Whether this is the protected anchor "home" zone. */
  is_default: boolean;
}
